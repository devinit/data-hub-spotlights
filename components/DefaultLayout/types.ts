export interface DefaultLayoutData {
  navigation: Navigation;
}

export interface Navigation {
  primary: NavigationItem[];
  secondary: NavigationItem[];
}

interface NavigationItem {
  title: string;
  full_url: string;
  active: boolean;
  relative_url: string;
}
