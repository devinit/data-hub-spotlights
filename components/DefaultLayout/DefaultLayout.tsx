import { NextComponentType } from 'next';
import React, { ReactNode, cloneElement, isValidElement, useState } from 'react';
import Header from '../Header/Header';
import { DefaultLayoutData } from './types';

const DefaultLayout: NextComponentType = ({ children }) => {
  const [ data, setData ] = useState<DefaultLayoutData | undefined>();

  const attachDataProp = (component: ReactNode) => {
    if (isValidElement(component)) {
      return cloneElement(component, { setData });
    }

    return component;
  };

  return (
    <div className="ui-base">
      <Header navigation={ data && data.navigation }/>
      <main id="pagecontent" className="pagecontent -nofocus" role="main" tabIndex={ -1 }>
        { attachDataProp(children) }
      </main>
    </div>
  );
};

export { DefaultLayout as default, DefaultLayout };
