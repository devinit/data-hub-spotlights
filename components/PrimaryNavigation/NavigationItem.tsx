import React, { FunctionComponent } from 'react';
import classNames from 'classnames';

export interface NavigationItemProps {
  title: string;
  active: boolean;
  url: string;
}

const NavigationItem: FunctionComponent<NavigationItemProps> = ({ active, title, url }) => {
  return (
    <li className={classNames('navigation-primary__item', { 'navigation-primary__item--active': active })}>
      <a href={url}>{title}</a>
    </li>
  );
};

export { NavigationItem };
