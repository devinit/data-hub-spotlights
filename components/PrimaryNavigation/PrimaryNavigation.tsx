import React, { FunctionComponent, ReactNode, useState } from 'react';
import classNames from 'classnames';
import { NavigationItem } from '../DefaultLayout';
import { NavigationItem as NavigationListItem } from './NavigationItem';

export interface PrimaryNavigationProps {
  items: NavigationItem[];
}

const PrimaryNavigation: FunctionComponent<PrimaryNavigationProps> = ({ items }) => {
  const [showMenu, setShowMenu] = useState(false);
  const renderNavigationItems = (): ReactNode =>
    items.map((item) => (
      <NavigationListItem key={item.title} title={item.title} active={item.active} url={item.full_url} />
    ));

  const onToggleMenu = (): void => {
    setShowMenu(!showMenu);
  };

  return (
    <>
      <button
        role="button"
        className={classNames('navigation-primary-toggle', { 'navigation-primary-toggle--active': showMenu })}
        id="navigation-primary-toggle"
        data-testid="navigation-primary-toggle"
        onClick={onToggleMenu}
      >
        <span>Menu</span>
      </button>
      <nav
        className={classNames('navigation-primary', { 'navigation-primary--active': showMenu })}
        role="navigation"
        id="navigation-primary"
        data-testid="navigation-primary"
      >
        <ul className="navigation-primary__items" data-testid="navigation-primary-ul">
          {renderNavigationItems()}
        </ul>
      </nav>
    </>
  );
};

export { PrimaryNavigation };
