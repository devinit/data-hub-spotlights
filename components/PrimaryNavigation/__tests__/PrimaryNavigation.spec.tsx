/**
 * @jest-environment jsdom
 */
/* eslint-disable @typescript-eslint/camelcase */
import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { PrimaryNavigation } from '../PrimaryNavigation';
import { NavigationItem } from '../../DefaultLayout';

const items: NavigationItem[] = [
  {
    title: 'Nav 1',
    full_url: 'https://localhost:3000/nav-1',
    active: false,
    relative_url: '/nav-1'
  },
  {
    title: 'Nav 2',
    full_url: 'https://localhost:3000/nav-2',
    active: true,
    relative_url: '/nav-2'
  }
];

describe('PrimaryNavigation', () => {
  test('renders correctly', () => {
    const renderer = TestRenderer.create(<PrimaryNavigation items={items} />).toJSON();

    expect(renderer).toMatchSnapshot();
  });

  test('adds & removes necessary classes on button toggle', () => {
    const { getByTestId } = render(<PrimaryNavigation items={items} />);

    const button = getByTestId('navigation-primary-toggle');
    const nav = getByTestId('navigation-primary');

    expect(button).not.toHaveClass('navigation-primary-toggle--active');
    expect(nav).not.toHaveClass('navigation-primary--active');

    fireEvent.click(button);

    expect(button).toHaveClass('navigation-primary-toggle--active');
    expect(nav).toHaveClass('navigation-primary--active');

    fireEvent.click(button);

    expect(button).not.toHaveClass('navigation-primary-toggle--active');
    expect(nav).not.toHaveClass('navigation-primary--active');
  });
});
