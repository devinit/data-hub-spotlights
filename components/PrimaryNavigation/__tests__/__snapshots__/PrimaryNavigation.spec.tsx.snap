// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryNavigation renders correctly 1`] = `
Array [
  <button
    className="navigation-primary-toggle"
    id="navigation-primary-toggle"
    onClick={[Function]}
    role="button"
  >
    <span>
      Menu
    </span>
  </button>,
  <nav
    className="navigation-primary"
    id="navigation-primary"
    role="navigation"
  >
    <ul
      className="navigation-primary__items"
      data-testid="navigation-primary-ul"
    >
      <li
        className="navigation-primary__item"
      >
        <a
          href="https://localhost:3000/nav-1"
        >
          Nav 1
        </a>
      </li>
      <li
        className="navigation-primary__item navigation-primary__item--active"
      >
        <a
          href="https://localhost:3000/nav-2"
        >
          Nav 2
        </a>
      </li>
    </ul>
  </nav>,
]
`;
